### PR TITLE
feat(organize_import): utilities for ordering import sources

### PR DESCRIPTION
**Pull request by Conaclos**
## Summary

Implements #4239.

In the RFC (#3015), I proposed to parse import sources in order to compare them in a structural way.
At the time of writing the RFC it took a lot of time trying to decide which structure to parse.
I left the RFC a bit loose because I couldn't find a satisfactory structure.

I fell into the same problem again. I am not 100% happy with the result and I am open to suggestions.
However, the current implementation should be good enough, and we can improve it later.

The implementation is in `biome_js_analyze/src/assists/source/organize_imports/util.rs` and `biome_string_case/src/lib.rs`.

In contrast to the spec, I decided to avoid a parsing phase (except for paths).
Instead, we just look at the first bytes of an import source to determine its **kind**.
An import source can be one of the following kind:

- _Unknown_: all import paths without a recognizable kind
- _Uri_, e.g. `https://example.org`
- _ProtocolPackage_, e.g. `node:test`, `npm:@scope/package`
- _Package_, e.g. `package`
- _Alias_: Node.js subpath imports and TypeScript aliases.
   This corresponds to import sources that start with `@/`, `#`, `~`, or `%`
    /// Import sources that start with `/`, `./`, or `../`
- _Path_: import sources that start with `/`, `./`, or `../`

Import paths are first ordered according to their kind (order of kind is reflected by their position in the previous list).
Note that unrecognized import source kinds are placed first.
I am unsure if we should keep or remove `Unknown` and choose a kind instead (probably Alias).

Two import sources with identical kind are then ordered by a custom order:
- Order between paths follows the proximity order (`../../` < `../` < `./`)
- Other kinds use a string collation (`ImportSourceAsciiCollator`) tailored for paths and URIs.
  The collation is case-insensitive (`A` < `a` < `B` < `b`), compares numbers in a human-friendly way (`9` < `10`).
  Special characters such as `/` and `@` are carefully ordered to yield a more natural ordering for import sources.

I added the base of the collation algorithm in `biome_string_case` because it is kind of related to character case.
Another option is to create a dedicated crate `biome_string_collator` or a crate `biome_str_utils`.

Note: This PR can be merged into `main` as the utilities introduced are not yet in use.

## Alternative design

Instead of creating a custom collation, we could use the conventional Unicode Collation reduced to the ASCII set (`biome_string_case::CldrAsciiCollator`) and to parse import sources in a more advanced structure. I started with this design in mind and find it a bit too complex. However, I am still open to revisiting it if you think it could be better.

## Test Plan

I added tests in the source code.

### CODE REVIEW

File: crates/biome_string_case/src/lib.rs [hunk 1]
```diff
@@ -415,25 +419,207 @@ const LEADING_BIT_INDEX_TO_CASE: [Case; 11] = [
...
+                }
+                (Some(_), None) => {
+                    return Ordering::Greater;
+                }
+            }
+        }
+    }
+}
+
+/// An [AsciiCollator] defines an order between ASCII characters.
```
Comment (ematipico) [inline crates/biome_string_case/src/lib.rs:515]:
```suggestion
/// An ASCII Collator defines an order between ASCII characters.
```

I am not sure it makes sense to document a type and add a link to itself

File: crates/biome_js_analyze/src/assists/source/organize_imports/util.rs [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+    }
+}
+
+/// Custom collation order to get a natural irder between import sources.
+///
+/// Non-printable characters and alphanumeric characters have the same order as [biome_string_case::CldrAsciiCollator].
+/// Others are ordered differently, but still between non-printable characters and alphanumeric characters.
+pub struct ImportSourceAsciiCollator;
+impl ImportSourceAsciiCollator {
+    const COLLATION: [u8; 128] = [
```
Comment (ematipico) [inline crates/biome_js_analyze/src/assists/source/organize_imports/util.rs:274]: Isn't this table the same we have in the other file?

File: crates/biome_string_case/src/lib.rs [hunk 3]
```diff
@@ -415,25 +419,207 @@ const LEADING_BIT_INDEX_TO_CASE: [Case; 11] = [
...
+        Self: Collator<Char = u8>,
+    {
+        self.cmp(s1.iter().copied(), s2.iter().copied())
+    }
+}
+impl<C: AsciiCollator> Collator for C {
+    type Char = u8;
+
+    fn weight(&self, c: &Self::Char) -> impl Ord {
+        unsafe { *Self::WEIGHTS.get_unchecked(*c as usize) }
```
Comment (ematipico) [inline crates/biome_string_case/src/lib.rs:555]: We should add a `// SAFETY` comment

File: crates/biome_string_case/src/lib.rs [hunk 4]
```diff
@@ -415,25 +419,207 @@ const LEADING_BIT_INDEX_TO_CASE: [Case; 11] = [
...
-    fn to_lowercase_cow(&self) -> std::borrow::Cow<Self>;
-    /// Returns the same value as String::to_lowercase. The only difference
-    /// is that this functions returns ```Cow``` and does not allocate
-    /// if the string is already in lowercase.
-    fn to_ascii_lowercase_cow(&self) -> std::borrow::Cow<Self>;
+    fn to_lowercase_cow(&self) -> Cow<Self>;
 }
 
-impl StrOnlyExtension for str {
+impl StrLikeExtension for str {
```
Comment (ematipico) [inline crates/biome_string_case/src/lib.rs:622]: Why did we rename the type? It seems unnecessary for the scope of the PR

File: crates/biome_string_case/src/lib.rs [hunk 5]
```diff
@@ -444,6 +630,12 @@ impl StrOnlyExtension for str {
         }
     }
 
+    fn ascii_nat_cmp(&self, other: &Self) -> Ordering {
```
Comment (ematipico) [inline crates/biome_string_case/src/lib.rs:633]: Docs missing

File: crates/biome_string_case/src/lib.rs [hunk 6]
```diff
@@ -399,7 +403,7 @@ impl Iterator for CasesIterator {
         (0, Some(6))
     }
 }
-impl std::iter::FusedIterator for CasesIterator {}
+impl core::iter::FusedIterator for CasesIterator {}
```
Comment (ematipico) [inline crates/biome_string_case/src/lib.rs:406]: Why the change?

Comment (Conaclos) [reply crates/biome_string_case/src/lib.rs:406]: No particular reason. I think using `core` is better than `std`. However, this should be moved to a separate PR. I will remove the change.

File: crates/biome_string_case/src/lib.rs (hunk 4 above)
Comment (Conaclos) [reply crates/biome_string_case/src/lib.rs:622]: I didn't rename the type, I just removed `to_ascii_lowercase_cow` from `StrOnlyExtension` because it is already present in `StrLikeExtension`. Now `str` implements both `StrLikeExtension` and `StrOnlyExtension`. This avoids duplicating all methods from `StrLikeExtension` in `StrOnlyExtension`.

File: crates/biome_string_case/src/lib.rs (hunk 5 above)
Comment (Conaclos) [reply crates/biome_string_case/src/lib.rs:633]: The docs is inherited. OI don't think we should repeat it?

File: crates/biome_js_analyze/src/assists/source/organize_imports/util.rs (hunk 2 above)
Comment (Conaclos) [reply crates/biome_js_analyze/src/assists/source/organize_imports/util.rs:274]: Another approach is to compare two import sources with the same kind as paths and to use the natural order defined in `biome_string_case`. It is less natural than the one I created in here but should be good for most of the cases.

File: crates/biome_string_case/src/lib.rs (hunk 5 above)
Comment (ematipico) [reply crates/biome_string_case/src/lib.rs:636]: My mistake :)

File: crates/biome_string_case/src/lib.rs (hunk 4 above)
Comment (ematipico) [reply crates/biome_string_case/src/lib.rs:625]: Thank you for the explanation, I just wanted to know if they were necessary or not, in order to reduce the diffs and make the PR easier to review. No worries.

File: crates/biome_js_analyze/src/assists/source/organize_imports/util.rs (hunk 2 above)
Comment (Conaclos) [reply crates/biome_js_analyze/src/assists/source/organize_imports/util.rs:274]: The [Unicode Collation Algorithm](https://www.unicode.org/reports/tr10/) allows ordering ain a more natural way characters.
The algorithm is a general and must respect some conformance rules.
It provides a default collation table named DUCET (Default Unicode Collation Element Table) that can be used as a base for any collation. It is expected that a locale defines a custom collation order for some characters and uses DUCET for the ones it doesn't care.

It seems that DUCET has been supplanted by a new collation table named [CLDR](https://en.wikipedia.org/wiki/Common_Locale_Data_Repository) (Common Locale Data Repository). DUCET and CLDR are mostly identical. In the ASCII side, only the order between punctuation and symbols is different between DUCET and CLDR. Nowadays, most of modern implementation relies on CLDR as a basic collation.

`CldrAsciiCollator` is CLDR reduced to the ASCII characters and extended to any byte by mapping all bytes higher than `0x7f` to themselves.

`ImportSourceAsciiCollator` is basically `CldrAsciiCollator` but with a different order for punctuation and symbols.
Let's take an example to highlight the differences between `CldrAsciiCollator` and `ImportSourceAsciiCollator`.
Given the following list of strings (sorted according to their binary representation):

```
/a-b/
/a/b/
https://example.org/path?prop-1=val
https://example.org/path?prop=val
npm:@scope/package@^10.0/path
npm:@scope/package@>=10.0/path
npm:@scope/package@10.0/path
npm:@scope/package@9.0/path
npm:@scope/package/path
```

Using `CldrAsciiCollator`, we obtain the following sorted list:

```
/a-b/
/a/b/
https://example.org/path?prop-1=val
https://example.org/path?prop=val
npm:@scope/package@^10.0/path
npm:@scope/package@>=10.0/path
npm:@scope/package@9.0/path
npm:@scope/package@10.0/path
npm:@scope/package/path
```

Using `ImportSourceAsciiCollator`, we obtain the following sorted list:

```
/a/b/
/a-b/
https://example.org/path?prop=val
https://example.org/path?prop-1=val
npm:@scope/package/path
npm:@scope/package@>=10.0/path
npm:@scope/package@^10.0/path
npm:@scope/package@9.0/path
npm:@scope/package@10.0/path
```

`ImportSourceAsciiCollator` is designed to provide a natural ordering of paths and URLs.
As you can note, `/a/b` is placed before `/a-b/` because `/` < `-` (while `-` < `/` in CLDR).
We have ordering of semantic specifier (`>` < `^` < `~`).
We are URL-friendly (`?` < `#` < `=` < `&`) that allows a good order between URLs with query and/or fragment parts.

---

I am thinking about comparing all imports of a same kind as paths. i.e. we first decompose an import into path components. This will ensure that `/` has the priority. If we make this change, we could reduce our dependency on `ImportSourceAsciiCollator` or even drop it and directly use `CldrAsciiCollator`.

Comment (ematipico) [reply crates/biome_js_analyze/src/assists/source/organize_imports/util.rs:274]: > Note: for now no code is using CldrAsciiCollator. It is here for potential future uses and to provide an example to how to implement AsciiCollator.

Thank you for the detailed explanation. I don't see any particular value to add code that won't be used (the `CldrAsciiCollator` won't be). I can't see any future usage as well.

**Review by ematipico — APPROVED** [review]

File: crates/biome_js_analyze/src/assists/source/organize_imports/util.rs (hunk 2 above)
Comment (Conaclos) [reply crates/biome_js_analyze/src/assists/source/organize_imports/util.rs:274]: > I can't see any future usage as well.

Actually I see a usage where we actually use natord (for example `useSortedKeys`).

Comment (ematipico) [reply crates/biome_js_analyze/src/assists/source/organize_imports/util.rs:274]: Awesome, thank you! Happy to merge the PR
